common/src/main/java/io/druid/math/expr/Parser.java

```diff
@@ -43,55 +42,44 @@
 public class Parser
 {
   private static final Logger log = new Logger(Parser.class);
-  private static final Map<String, Supplier<Function>> func;
+  private static final Map<String, Function> FUNCTIONS;
 
   static {
-    Map<String, Supplier<Function>> functionMap = Maps.newHashMap();
+    Map<String, Function> functionMap = Maps.newHashMap();
     for (Class clazz : Function.class.getClasses()) {
       if (!Modifier.isAbstract(clazz.getModifiers()) && Function.class.isAssignableFrom(clazz)) {
         try {
-          Function function = (Function)clazz.newInstance();
-          if (function instanceof Function.FunctionFactory) {
-            functionMap.put(function.name().toLowerCase(), (Supplier<Function>) function);
-          } else {
-            functionMap.put(function.name().toLowerCase(), Suppliers.ofInstance(function));
-          }
+          Function function = (Function) clazz.newInstance();
+          functionMap.put(function.name().toLowerCase(), function);
         }
         catch (Exception e) {
           log.info("failed to instantiate " + clazz.getName() + ".. ignoring", e);
         }
       }
     }
-    func = ImmutableMap.copyOf(functionMap);
-  }
-
-  public static Function getFunction(String name) {
-    Supplier<Function> supplier = func.get(name.toLowerCase());
-    if (supplier == null) {
-      throw new IAE("Invalid function name '%s'", name);
-    }
-    return supplier.get();
+    FUNCTIONS = ImmutableMap.copyOf(functionMap);
   }
 
-  public static boolean hasFunction(String name)
+  public static Function getFunction(String name)
```

Comment:
Yes, I think at some point, the expressions should be more aware of types. It will help with performance too. I think in the future, we want to be doing code generation and that will require advance knowledge of what type everything is going to be.

For now, they aren't type-aware.